src/PackagePrivate/DatabaseTermIdsResolver.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Wikibase\TermStore\MediaWiki\PackagePrivate;
+
+use InvalidArgumentException;
+use stdClass;
+use Wikimedia\Rdbms\IDatabase;
+use Wikimedia\Rdbms\ILoadBalancer;
+use Wikimedia\Rdbms\IResultWrapper;
+
+/**
+ * Term ID resolver using the normalized database schema.
+ *
+ * @license GPL-2.0-or-later
+ */
+class DatabaseTermIdsResolver implements TermIdsResolver {
+
+	/** @var TypeIdsResolver */
+	private $typeIdsResolver;
+
+	/** @var ILoadBalancer */
+	private $lb;
+
+	/** @var IDatabase */
```

Comment:
IDatabase|null

Reply:
I marked it non-null intentionally, because most of the class doesn’t need to worry that the database might be null: after `connectDbr`/`connectDbw` runs, the rest of the code can just assume that the connections are populated.